### ISSUE TITLE
[Bug]: Strings cut - unplanned/undocumented behaviours

### ISSUE THREAD

**Issue by dave-csc** [apache/hop#4624]
### Apache Hop version?

2.10.0

### Java version?

17.0.2

### Operating system

Linux

### What happened?

Undocumented behaviours:
- Set Cut from = 0 or lower, Cut to = a value greater than the string length: the entire string will be returned
- Set Cut from = greater than 0, Cut to = a value greater than the string length: the substring from the "Cut from" index (+1) to the end of the original string will be returned
- Set Cut from = 0, Cut to = a negative number (e.g. -10): the substring composed of the last (10) characters will be returned

Unexpected behaviour:
- Set Cut from and Cut to both negative (e.g. -10 and -5): NULL will be returned, even if Cut to is greater (lower absolute value) than Cut from -> probably a better choice should be returning the substring from length - (10) to length - (5)

### Issue Priority

Priority: 3

### Issue Component

Component: Documentation, Component: Transforms

**Comment by hansva** [comment]
The cutString logic:

```
private String cutString(String string, int cutFrom, int cutTo) {
    String rcode = string;

    if (!Utils.isEmpty(rcode)) {
      int lenCode = rcode.length();

      if ((cutFrom >= 0 && cutTo >= 0) && cutFrom > lenCode) {
        rcode = null;
      } else if ((cutFrom >= 0 && cutTo >= 0) && (cutTo < cutFrom)) {
        rcode = null;
      } else if ((cutFrom < 0 && cutTo < 0) && cutFrom < -lenCode) {
        rcode = null;
      } else if ((cutFrom < 0 && cutTo < 0) && (cutFrom < cutTo)) {
        rcode = null;
      } else {
        if (cutTo > lenCode) {
          cutTo = lenCode;
        }
        if (cutTo < 0 && cutFrom == 0 && (-cutTo) > lenCode) {
          cutTo = -(lenCode);
        }
        if (cutTo < 0 && cutFrom < 0 && (-cutTo) > lenCode) {
          cutTo = -(lenCode);
        }

        if (cutFrom >= 0 && cutTo > 0) {
          rcode = rcode.substring(cutFrom, cutTo);
        } else if (cutFrom < 0 && cutTo < 0) {
          rcode = rcode.substring(rcode.length() + cutTo, lenCode + cutFrom);
        } else if (cutFrom == 0 && cutTo < 0) {
          int intFrom = rcode.length() + cutTo;
          rcode = rcode.substring(intFrom, lenCode);
        }
      }
    }

    return rcode;
  }
```

**Comment by hansva** [comment]
After thinking about it and looking at the code, that flip would cause extra complexity to the code for something that actually works.

It would also cause a regression issue for the people who are already using it. I originally thought the code did not allow 2 negative numbers. The explanation could be as simple as the `from` location should always be closest to 0

**Comment by dave-csc** [comment]
Added some examples in the docs (see [pull request](https://github.com/apache/hop/pull/5565)).

All the examples have been tested with Hop 2.14: anyway, I still find some of them kinda weird... 🤪

**Comment by hansva** [comment]
Yeah, discussions could be had about some of the returned results and null vs empty string.
But having the current behaviour documented is a great step forward!

thanks for the work @dave-csc
